### DISCUSSION TITLE
sign BIP49 P2SH-P2WPKH transaction for litecoin

### DISCUSSION THREAD

**Discussion by 3tbraden** [trustwallet/wallet-core#1392] [Q&A]
I took reference from the example here
https://github.com/trustwallet/wallet-core/blob/9c5927e5ac45fb57339f4d91efbc71154aaa2c70/swift/Tests/Blockchains/BItcoinTests.swift#L84

except that I tried doing the steps with litecoin and input address such as `MTt3sGWkFQQVhFh11Jk6bxSWv5C6VamAdu`, but generating the script with
```
p2pkhScript := C.TWBitcoinScriptBuildPayToWitnessPubkeyHash(pubkeyhash)
p2pkhScriptData := C.TWBitcoinScriptData(p2pkhScript)
script := TWDataGoBytes(C.TWHashSHA256RIPEMD(p2pkhScriptData))
signInput.utxo.script = script
```
and `AnySigner.Sign` would return `error:"Invalid output script"`;
while trying with
```
script := []byte{}
script = append(script, 0x00, 0x00)
script[1] = 0x14
script = append(script, pubkeyhash...)
signInput.utxo.script = script
```
it signs and decodes the transaction without problem, but when trying to broadcast the raw tx in hex on blockchair.com, I got `bad-txns-nonstandard-inputs (code 64)`

I have also tried another approach following the spec: https://github.com/bitcoin/bips/blob/master/bip-0049.mediawiki
```
scriptSig := []byte{}
scriptSig = append(script, 0x00, 0x00)
scriptSig[1] = 0x14
scriptSig = append(script, pubkeyhash...)

script = []byte{0xa9, 0x14}
script = append(script, TWDataGoBytes(C.TWHashSHA256RIPEMD(scriptSig))...)
script = append(script, 0x87)
```
but then I got `error:"Missing redeem script`

I am using go and I hope the above example provides a clear illustration. Could you advise which part have I got wrong or is nested segwit for litecoin supported in wallet-core?

**Comment by optout21** [comment]
This address is not a witness address, so PayToWitnessPubkeyHash will not work with it.  Use segwit address, starting with "ltc...".

**Reply by 3tbraden** [reply]
this is a BIP49 P2SH-P2WPKH address, as suggested in https://github.com/trustwallet/wallet-core/issues/812 I should use `buildPayToWitnessPubKeyHash()`?

**Comment by 3tbraden** [comment]
could it be p2sh addresses in litecoin are not correctly handled? https://github.com/trustwallet/wallet-core/blob/76c53ec6faf65d94a60e13c4b44b0301756af76b/coins.json#L44 here it shows prefix is `0x50` but litecoin p2sh addresses now begins with `0x32`: https://bitcoin.stackexchange.com/a/80355/85343

**Reply by optout21** [reply]
0x32 is 50 :)

**Comment by 3tbraden** [comment]
I have tried another approach following the script injection in the test: https://github.com/trustwallet/wallet-core/blob/9c5927e5ac45fb57339f4d91efbc71154aaa2c70/swift/Tests/Blockchains/BItcoinTests.swift#L112-L114
but got `error:"Missing redeem script."` trying to sign.

I tried the following:
```go
for i, _ := signingInput.Utxo {
  ...
  p2shScript := C.TWBitcoinScriptLockScriptForAddress("MTt3sGWkFQQVhFh11Jk6bxSWv5C6VamAdu", uint32(2))
  p2shScriptData := C.TWBitcoinScriptData(p2shScript)

  pubkeyHash := C.TWHashSHA256RIPEMD(C.TWPublicKeyData(pub))
  p2wpkhScript := C.TWBitcoinScriptBuildPayToWitnessPubkeyHash(pubkeyHash)
  p2wpkhScriptData := C.TWBitcoinScriptData(p2wpkhScript)

  signingInput.Scripts[hex.EncodeToString(TWDataGoBytes(p2shScriptData))] = p2wpkhScriptData
  signingInput.Utxo[i].Script = p2shScriptData
}
```
If I use `TWBitcoinScriptMatchPayToWitnessPublicKeyHash` and `TWBitcoinScriptMatchPayToScriptHash` to extract script data instead of `TWBitcoinScriptData`, I would get `error: Invalid output script` instead.
